### PR TITLE
preserve node stacktraces from compiled autograd through AOTDispatcher, due to GmWrapper

### PR DESCRIPTION
**Pull request by bdhirsh**
Fixes https://github.com/pytorch/pytorch/issues/133567

New log output from the repro:
```
(/home/hirsheybar/local/b/pytorch-env) [hirsheybar@devgpu001.lla3 ~/local/b/pytorch (compiled_autograd_stacktraces)]$ TORCH_LOGS="compiled_autograd_verbose,aot" python tmp5.py
INFO: TRACED GRAPH
 ===== Joint graph 0 =====
 /home/hirsheybar/local/b/pytorch/torch/fx/_lazy_graph_module.py class joint_helper(torch.nn.Module):
    def forward(self, primals, tangents):
        primals_1: "f32[4, 4][4, 1]cpu"; tangents_1: "f32[4, 4][4, 1]cpu";

        primals_1, tangents_1, = fx_pytree.tree_flatten_spec([primals, tangents], self._in_spec)
         # File: /home/hirsheybar/local/b/pytorch/tmp5.py:6 in f, code: return torch.matmul(x, x)
        mm: "f32[4, 4][4, 1]cpu" = torch.ops.aten.mm.default(primals_1, primals_1)
        permute: "f32[4, 4][1, 4]cpu" = torch.ops.aten.permute.default(primals_1, [1, 0])
        mm_1: "f32[4, 4][4, 1]cpu" = torch.ops.aten.mm.default(permute, tangents_1);  permute = None
        permute_1: "f32[4, 4][1, 4]cpu" = torch.ops.aten.permute.default(primals_1, [1, 0]);  primals_1 = None
        mm_2: "f32[4, 4][4, 1]cpu" = torch.ops.aten.mm.default(tangents_1, permute_1);  tangents_1 = permute_1 = None

         # File: /home/hirsheybar/local/b/pytorch/tmp5.py:6 in f, code: return torch.matmul(x, x)
        add: "f32[4, 4][4, 1]cpu" = torch.ops.aten.add.Tensor(mm_2, mm_1);  mm_2 = mm_1 = None
        return pytree.tree_unflatten([mm, add], self._out_spec)


INFO: aot_config id: 0, fw_metadata=ViewAndMutationMeta(input_info=[InputAliasInfo(is_leaf=True, mutates_data=False, mutates_metadata=False, mutations_hidden_from_autograd=True, mutations_under_no_grad_or_inference_mode=False, mutation_inductor_storage_resize=False, mutates_storage_metadata=False, requires_grad=True, keep_input_mutations=False)], output_info=[OutputAliasInfo(output_type=<OutputType.non_alias: 1>, raw_type=<class 'torch._subclasses.functional_tensor.FunctionalTensor'>, base_idx=None, dynamic_dims=set(), requires_grad=True, functional_tensor=None)], num_intermediate_bases=0, keep_input_mutations=False, traced_tangents=[FakeTensor(..., size=(4, 4))], subclass_inp_meta=[0], subclass_fw_graph_out_meta=[0], subclass_tangent_meta=[0], is_train=True, traced_tangent_metas=None, num_symints_saved_for_bw=0, grad_enabled_mutation=None, deterministic=False, static_input_indices=[], tokens={}, indices_of_inputs_that_requires_grad_with_mutations_in_bw=[], bw_donated_idxs=None), inner_meta=ViewAndMutationMeta(input_info=[InputAliasInfo(is_leaf=True, mutates_data=False, mutates_metadata=False, mutations_hidden_from_autograd=True, mutations_under_no_grad_or_inference_mode=False, mutation_inductor_storage_resize=False, mutates_storage_metadata=False, requires_grad=True, keep_input_mutations=False)], output_info=[OutputAliasInfo(output_type=<OutputType.non_alias: 1>, raw_type=<class 'torch._subclasses.functional_tensor.FunctionalTensor'>, base_idx=None, dynamic_dims=set(), requires_grad=True, functional_tensor=None)], num_intermediate_bases=0, keep_input_mutations=False, traced_tangents=[FakeTensor(..., size=(4, 4))], subclass_inp_meta=[0], subclass_fw_graph_out_meta=[0], subclass_tangent_meta=[0], is_train=True, traced_tangent_metas=None, num_symints_saved_for_bw=0, grad_enabled_mutation=None, deterministic=False, static_input_indices=[], tokens={}, indices_of_inputs_that_requires_grad_with_mutations_in_bw=[], bw_donated_idxs=None)
INFO: TRACED GRAPH
 ===== Forward graph 0 =====
 /home/hirsheybar/local/b/pytorch/torch/fx/_lazy_graph_module.py class GraphModule(torch.nn.Module):
    def forward(self, primals_1: "f32[4, 4][4, 1]cpu"):
         # File: /home/hirsheybar/local/b/pytorch/tmp5.py:6 in f, code: return torch.matmul(x, x)
        mm: "f32[4, 4][4, 1]cpu" = torch.ops.aten.mm.default(primals_1, primals_1)
        permute: "f32[4, 4][1, 4]cpu" = torch.ops.aten.permute.default(primals_1, [1, 0]);  primals_1 = None
        return (mm, permute)


INFO: TRACED GRAPH
 ===== Backward graph 0 =====
 <eval_with_key>.1 class GraphModule(torch.nn.Module):
    def forward(self, permute: "f32[4, 4][1, 4]cpu", tangents_1: "f32[4, 4][4, 1]cpu"):
         # File: /home/hirsheybar/local/b/pytorch/tmp5.py:6 in f, code: return torch.matmul(x, x)
        mm_1: "f32[4, 4][4, 1]cpu" = torch.ops.aten.mm.default(permute, tangents_1)
        mm_2: "f32[4, 4][4, 1]cpu" = torch.ops.aten.mm.default(tangents_1, permute);  tangents_1 = permute = None

         # File: /home/hirsheybar/local/b/pytorch/tmp5.py:6 in f, code: return torch.matmul(x, x)
        add: "f32[4, 4][4, 1]cpu" = torch.ops.aten.add.Tensor(mm_2, mm_1);  mm_2 = mm_1 = None
        return (add,)


DEBUG: Cache miss due to new autograd node: torch::autograd::GraphRoot (NodeCall 0) with key size 39, previous key sizes=[]
DEBUG: TRACED GRAPH
 ===== Compiled autograd graph =====
 <eval_with_key>.2 class CompiledAutograd(torch.nn.Module):
    def forward(self, inputs, sizes, scalars, hooks):
        # No stacktrace found for following nodes
        getitem: "f32[]cpu" = inputs[0]
        getitem_1: "f32[4, 4]cpu" = inputs[1]
        getitem_2: "f32[4, 4]cpu" = inputs[2];  inputs = None

         # File: /home/hirsheybar/local/b/pytorch/torch/_dynamo/compiled_autograd.py:379 in set_node_origin, code: SumBackward0 (NodeCall 1)
        expand: "f32[4, 4]cpu" = torch.ops.aten.expand.default(getitem, [4, 4]);  getitem = None

         # File: /home/hirsheybar/local/b/pytorch/torch/_dynamo/compiled_autograd.py:379 in set_node_origin, code: CompiledFunctionBackward (NodeCall 2)
        clone: "f32[4, 4]cpu" = torch.ops.aten.clone.default(expand, memory_format = torch.contiguous_format);  expand = None
        mm: "f32[4, 4]cpu" = torch.ops.aten.mm.default(getitem_1, clone)
        mm_1: "f32[4, 4]cpu" = torch.ops.aten.mm.default(clone, getitem_1);  clone = getitem_1 = None
        add: "f32[4, 4]cpu" = torch.ops.aten.add.Tensor(mm_1, mm);  mm_1 = mm = None

         # File: /home/hirsheybar/local/b/pytorch/torch/_dynamo/compiled_autograd.py:379 in set_node_origin, code: torch::autograd::AccumulateGrad (NodeCall 3)
        accumulate_grad_ = torch.ops.inductor.accumulate_grad_.default(getitem_2, add);  getitem_2 = add = accumulate_grad_ = None
        _exec_final_callbacks_stub = torch__dynamo_external_utils__exec_final_callbacks_stub();  _exec_final_callbacks_stub = None
        return []


INFO: TRACED GRAPH
 ===== Forward graph 1 =====
 /home/hirsheybar/local/b/pytorch/torch/fx/_lazy_graph_module.py class <lambda>(torch.nn.Module):
    def forward(self, arg0_1: "f32[][]cpu", arg1_1: "f32[4, 4][1, 4]cpu", arg2_1: "f32[4, 4][4, 1]cpu"):
         # File: /home/hirsheybar/local/b/pytorch/torch/_dynamo/compiled_autograd.py:379 in set_node_origin, code: SumBackward0 (NodeCall 1)
        expand: "f32[4, 4][0, 0]cpu" = torch.ops.aten.expand.default(arg0_1, [4, 4]);  arg0_1 = None

         # File: /home/hirsheybar/local/b/pytorch/torch/_dynamo/compiled_autograd.py:379 in set_node_origin, code: CompiledFunctionBackward (NodeCall 2)
        clone: "f32[4, 4][4, 1]cpu" = torch.ops.aten.clone.default(expand, memory_format = torch.contiguous_format);  expand = None
        mm: "f32[4, 4][4, 1]cpu" = torch.ops.aten.mm.default(arg1_1, clone)
        mm_1: "f32[4, 4][4, 1]cpu" = torch.ops.aten.mm.default(clone, arg1_1);  clone = arg1_1 = None
        add: "f32[4, 4][4, 1]cpu" = torch.ops.aten.add.Tensor(mm_1, mm);  mm_1 = mm = None

         # File: /home/hirsheybar/local/b/pytorch/torch/_dynamo/polyfill.py:44 in accumulate_grad, code: new_grad = torch.clone(new_grad)
        clone_1: "f32[4, 4][4, 1]cpu" = torch.ops.aten.clone.default(add);  add = None
        return (clone_1,)
```

The problem was that we expect the input to AOTAutograd to be a GraphModule in order to do all of the fancy stacktrace preservation logic, but we now need to handle compiled autograd passing in a `GmWrapper` instead (which it uses to try to preserve input boxing, so inductor can properly free activations)

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #133581
* #133580
* __->__ #133574

### CODE REVIEW

**Comment by bdhirsh** [comment]
CI is unhappy - from a quick look, I'm failing to ensure that the other args to the compiled backward (like hooks) are properly accounted for in the graph

File: torch/_functorch/_aot_autograd/traced_function_transforms.py [hunk 1]
```diff
@@ -792,21 +792,35 @@ def create_functional_call(mod, params_spec, params_len, store_orig_mod=False):
...
-                            *args[params_len:], **kwargs
-                        )
+                        if isinstance(mod, torch._dynamo.utils.GmWrapper):
+                            # Assumption : GmWrapper is only used by compiled autograd.
+                            # it stores no buffers/params and has no kwargs.
+                            # we need to carefully **not** unwrap the args list
+                            # to ensure args are boxed.
+                            assert params_len == 0
+                            assert len(kwargs) == 0
+                            out = PropagateUnbackedSymInts(mod_).run(args)
```
Comment (xmfan) [inline torch/_functorch/_aot_autograd/traced_function_transforms.py:817]: there's some logic in GmWrapper.forward that we'll need here: https://github.com/pytorch/pytorch/blob/90d2593b3e819002157cab90d92351fcfe3ac59b/torch/_dynamo/utils.py#L2907-L2909

File: torch/_functorch/_aot_autograd/traced_function_transforms.py [hunk 2]
```diff
@@ -792,21 +792,35 @@ def create_functional_call(mod, params_spec, params_len, store_orig_mod=False):
     # https://github.com/pytorch/pytorch/issues/103569
 
     def functional_call(*args, **kwargs):
+        nonlocal mod
```
Comment (ezyang) [inline torch/_functorch/_aot_autograd/traced_function_transforms.py:795]: Why nonlocal? Are you assigning over mod?

File: torch/_functorch/_aot_autograd/traced_function_transforms.py [hunk 3]
```diff
@@ -792,21 +792,35 @@ def create_functional_call(mod, params_spec, params_len, store_orig_mod=False):
...
+        # Need to unwrap and go down the fx.Interpreter.run() path below
+        # in order to properly preserve node stacktraces
+        mod_ = mod.gm if isinstance(mod, torch._dynamo.utils.GmWrapper) else mod
+
         with stateless._reparametrize_module(
-            mod, pytree.tree_unflatten(args[:params_len], params_spec)
+            mod_, pytree.tree_unflatten(args[:params_len], params_spec)
         ), maybe_disable_thunkify():
-            if isinstance(mod, torch.fx.GraphModule):
+            if isinstance(mod, (torch.fx.GraphModule, torch._dynamo.utils.GmWrapper)):
```
Comment (ezyang) [inline torch/_functorch/_aot_autograd/traced_function_transforms.py:803]: Why not do the test here on `mod_`?

File: torch/_functorch/_aot_autograd/traced_function_transforms.py (hunk 1 above)
Comment (ezyang) [reply torch/_functorch/_aot_autograd/traced_function_transforms.py:817]: You should probably just have a "middleware" wrapper that uniformly takes care of unwrapping GmWrapper and modifying the calling convention, should be cleaner.

@xmfan is there a reason we HAVE to have a GmWrapper? Shouldn't custom GraphModule prelude/postlude be enough here?

Comment (xmfan) [reply torch/_functorch/_aot_autograd/traced_function_transforms.py:817]: What is the prelude/postlude? We use GmWrapper to work around the dynamo GraphModule needing boxed inputs, but AOTDispatcher always tracing the GraphModule with flat inputs

Comment (ezyang) [reply torch/_functorch/_aot_autograd/traced_function_transforms.py:817]: Code generation for fx.Graph can be overridden via _codegen field. For example, this is used to generate GraphModule that can take arbitrary pytree as argument, it manages flattening/unflattening in the body. You could potentially use a similar mechanism to implement GmWrapper. cc @suo @Chillee

Comment (xmfan) [reply torch/_functorch/_aot_autograd/traced_function_transforms.py:817]: I can look into it, but we probably still need GmWrapper for non-dynamo frontends that are passing in non-overriden graphs

**Comment by yf225** [comment]
@bdhirsh I believe this would be super useful for compiled autograd debugging in general!

**Pull request closed without merging**